### ISSUE TITLE
Linux support

### ISSUE THREAD

**Issue by dromer** [DamRsn/NeuralNote#33]
Duh.

**Comment by DamRsn** [comment]
Contributions are welcome 🙂

The building process should be very similar to OSx.

**Comment by gnac** [comment]
I've made some steps to build this (and the /libonnxruntime-neuralnote dependency) in Linux. 
At the moment, I've got it NeuralNote compiling but it fails in the linker stage.  I'll  push what I have to my [NeuralNote](https://github.com/gnac/NeuralNote) and [onnxruntime lib](https://github.com/gnac/libonnxruntime-neuralnote) forks. But in short, I'm getting the following error:
``` 
[100%] Built target NeuralNote_Standalone
/usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: libBasicPitchCNN.a(BasicPitchCNN.cpp.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
```

**Comment by gnac** [comment]
I've created a (pull request](https://github.com/DamRsn/NeuralNote/pull/44) that is dependent on a prebuilt onnxruntime library.  It isn't building (linking errors) yet, but I figured its a place to start a discussion.  
FWIW, I also have a pending [pull request on the libonnx project](https://github.com/tiborvass/libonnxruntime-neuralnote/pull/1) which adds support for building the linux archive.  You will probably need to pull this project to build an archive to test the linux build of this pull request.

**Comment by RustoMCSpit** [comment]
just tried running this on wine with yabridge on bitwig, it kinda works but the gui has a weird rendering issues which causes huge delay / latency

**Comment by dromer** [comment]
@RustoMCSpit but we have the source-code, so no need for wine.

Thnx for the work @gnac not sure if I'll be able to get your stuff tested, but looking forward to see it working!

**Comment by RustoMCSpit** [comment]
> @RustoMCSpit but we have the source-code, so no need for wine.
> 
> Thnx for the work @gnac not sure if I'll be able to get your stuff tested, but looking forward to see it working!

no need for wine until it actually is supported you mean right?

**Comment by dromer** [comment]
@RustoMCSpit sure, if it works for you and you can play with the software that's great.

**Comment by RustoMCSpit** [comment]
> @RustoMCSpit sure, if it works for you and you can play with the software that's great.

ehhh, 'works' but barely, very bugged

**Comment by RustoMCSpit** [comment]
yeah after a few days the app is pretty unbareable to work with through WINE, if the devs want to see it through they could collect crash reports they get and see whats going on

**Comment by dromer** [comment]
@RustoMCSpit if you want to go that route please open a separate ticket so we don't mix up topics.

I think it would make more sense to build the program for Linux then to try and debug wine issues.

**Comment by thelabcat** [comment]
Lemme know if this goes anywhere! RN it is only usable in a VM for me ;-(

**Comment by RustoMCSpit** [comment]
just to update, program still really weird on WINE after new update

**Comment by dromer** [comment]
@RustoMCSpit please open a separate topic on WINE, this is request is about actual Linux builds and NOT any WINE work 
around.

Lets not dilute either topic by cramming it into the same ticket.

**Comment by polygon** [comment]
@gnac Did you get any further with this? I've also managed to build my own version of onnxruntime and actually got NeuralNote to compile. But right now it segfaults when starting. I've so far traced this back to this line of code (juce_XmlElement.cpp):

`static const String juce_xmltextContentAttributeName ("text");`

For some reason, this static variable is not initialized correctly and the pointer inside the string (to the actual text data) is a Null pointer so it crashes when it's being used (in this case, to parse the Linux fontconfig). Might just be a small issue with some JUCE settings required on Linux, I'm sifting through the things, but it's a pretty strange error.

**Comment by polygon** [comment]
To me, this looks like a case of "static initialization fiasco". In `NeuralNote/Lib/Components/UIDefines.h` there are pointers to fonts declared like so:

```
// Fonts
const juce::Typeface::Ptr MONTSERRAT_BOLD =
    juce::Typeface::createSystemTypefaceFor(BinaryData::MontserratBold_ttf, BinaryData::MontserratBold_ttfSize);

const juce::Typeface::Ptr MONTSERRAT_SEMIBOLD =
    juce::Typeface::createSystemTypefaceFor(BinaryData::MontserratSemiBold_ttf, BinaryData::MontserratSemiBold_ttfSize);

const juce::Typeface::Ptr MONTSERRAT_REGULAR =
    juce::Typeface::createSystemTypefaceFor(BinaryData::MontserratRegular_ttf, BinaryData::MontserratRegular_ttfSize);
```

Under Linux, this triggers (unless JUCE_FONT_PATH is set) a font lookup that involves parsing `/etc/fonts/fonts.conf` using the XML parser. The XML parser has the static string initialization mentioned above. In my compiled variant, this static string has not been initialized when the `createSystemTypefaceFor` calls are being made. I guess this works for Windows and Mac because loading fonts there does not involve the XML parser. Still, this could be considered unsound (at least on Linux). I will try to work around this.

Setting JUCE_FONT_PATH seems to solve the problem and makes the program start. However, so far I've not yet seen a window open. Still working on it.

**Comment by gnac** [comment]
@polygon you've gotten further than I have.  How did you resolve the linker issue?

**Comment by polygon** [comment]
I don't remember running into this linker issue to be honest. I've uploaded my current state here: https://github.com/polygon/NeuralNote/tree/linux

I re-factored the UIDefines.h code into a singleton provider to work around the initialization order issues. Setting JUCE_FONT_PATH would solve the issue for running the program standalone but not when used as a plugin. With the refactoring, I also got things working in Bitwig.

I'm not currently using `build.sh` but just run CMake directly like so:

```
cmake -B build -DCMAKE_BUILD_TYPE="Release"
cmake --build build -j
```

This expects to have a Linux built libonnxruntime-neuralnote in place as well, similarly unpacked as described in the `build.sh`. I've also got this done, but it's still quite hacky and I'm in the process of writing scripts for all of this. So you can poke around if you want, but it's still all a bit hacky and I'm working on consolidating everything.

**Comment by CodesoundR** [comment]
> I don't remember running into this linker issue to be honest. I've uploaded my current state here: https://github.com/polygon/NeuralNote/tree/linux
> 
> I re-factored the UIDefines.h code into a singleton provider to work around the initialization order issues. Setting JUCE_FONT_PATH would solve the issue for running the program standalone but not when used as a plugin. With the refactoring, I also got things working in Bitwig.
> 
> I'm not currently using `build.sh` but just run CMake directly like so:
> 
> ```
> cmake -B build -DCMAKE_BUILD_TYPE="Release"
> cmake --build build -j
> ```
> 
> This expects to have a Linux built libonnxruntime-neuralnote in place as well, similarly unpacked as described in the `build.sh`. I've also got this done, but it's still quite hacky and I'm in the process of writing scripts for all of this. So you can poke around if you want, but it's still all a bit hacky and I'm working on consolidating everything.

Hi, and thanks so much for porting this beautiful project in Linux (I'm under Mint).
I ask you if you have any news on the script you proposed. I'm having a hard time figuring out how to install this softare...
thanks again!!!

**Comment by Ken-Andre** [comment]
> > I don't remember running into this linker issue to be honest. I've uploaded my current state here: https://github.com/polygon/NeuralNote/tree/linux
> > I re-factored the UIDefines.h code into a singleton provider to work around the initialization order issues. Setting JUCE_FONT_PATH would solve the issue for running the program standalone but not when used as a plugin. With the refactoring, I also got things working in Bitwig.
> > I'm not currently using `build.sh` but just run CMake directly like so:
> > ```
> > cmake -B build -DCMAKE_BUILD_TYPE="Release"
> > cmake --build build -j
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This expects to have a Linux built libonnxruntime-neuralnote in place as well, similarly unpacked as described in the `build.sh`. I've also got this done, but it's still quite hacky and I'm in the process of writing scripts for all of this. So you can poke around if you want, but it's still all a bit hacky and I'm working on consolidating everything.
> 
> Hi, and thanks so much for porting this beautiful project in Linux (I'm under Mint). I ask you if you have any news on the script you proposed. I'm having a hard time figuring out how to install this softare... thanks again!!!


Yes it will be really great.
Think that they had already started working on it.

**Comment by CodesoundR** [comment]
> Yes it will be really great.
> Think that they had already started working on it.

Thanks .....  I'm waiting.... ;-)

**Comment by polygon** [comment]
I drove this a bit further and opened Pull Requests in both project components:

https://github.com/DamRsn/NeuralNote/pull/85
https://github.com/tiborvass/libonnxruntime-neuralnote/pull/2

However, communication quickly died down (or didn't happen), so I assumed that the maintainers were not interested in this. I finished the packaging for the Linux distribution of my choice (NixOS) and carried on. I'll give a short rundown on how to build this yourself, though. You will need to use my forks of [libonnxruntime-neuralnote](https://github.com/polygon/libonnxruntime-neuralnote) and [NeuralNote](https://github.com/polygon/NeuralNote). For both, you want the "Linux" branch.

I will link my Nix build recipes, since they are also a good indicator on the dependencies that you will need. You can ignore a lot of the CMake download links replacement, that is a Nix/NixOS peculiarity (no internet access during builds).

# libonnxruntime-neuralnote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/libonnxruntime-neuralnote.nix))

Once you installed everything unter [`buildInputs`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L32-L41), you just run the instructions under [`buildPhase` ](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L52-L55). This should yield you a `libonnxruntime-neuralnote.tar.gz` which you will need in the next step.

# NeuralNote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/neuralnote.nix))

Again, make sure you have all the [tools and dependencies](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L42-L58) installed. Now, run all the steps under [`postPatch`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L77-L86) in the root-folder of the repo. This also patches a Drag&Drop bug in JUCE. You then just run CMake like [above](https://github.com/DamRsn/NeuralNote/issues/33#issuecomment-1784055326) and this should give you the final binary / plugin.

Hopefully this gets you started. Let me know if you get stuck anywhere.

**Comment by thelabcat** [comment]
Thanks for doing this.

**Comment by CodesoundR** [comment]
> I drove this a bit further and opened Pull Requests in both project components:
> 
> #85 [tiborvass/libonnxruntime-neuralnote#2](https://github.com/tiborvass/libonnxruntime-neuralnote/pull/2)
> 
> However, communication quickly died down (or didn't happen), so I assumed that the maintainers were not interested in this. I finished the packaging for the Linux distribution of my choice (NixOS) and carried on. I'll give a short rundown on how to build this yourself, though. You will need to use my forks of [libonnxruntime-neuralnote](https://github.com/polygon/libonnxruntime-neuralnote) and [NeuralNote](https://github.com/polygon/NeuralNote). For both, you want the "Linux" branch.
> 
> I will link my Nix build recipes, since they are also a good indicator on the dependencies that you will need. You can ignore a lot of the CMake download links replacement, that is a Nix/NixOS peculiarity (no internet access during builds).
> # libonnxruntime-neuralnote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/libonnxruntime-neuralnote.nix))
> 
> Once you installed everything unter [`buildInputs`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L32-L41), you just run the instructions under [`buildPhase` ](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L52-L55). This should yield you a `libonnxruntime-neuralnote.tar.gz` which you will need in the next step.
> # NeuralNote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/neuralnote.nix))
> 
> Again, make sure you have all the [tools and dependencies](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L42-L58) installed. Now, run all the steps under [`postPatch`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L77-L86) in the root-folder of the repo (in line 83, reference the `tar.gz` file from the build above). This also patches a Drag&Drop bug in JUCE. You then just run CMake like [above](https://github.com/DamRsn/NeuralNote/issues/33#issuecomment-1784055326) and this should give you the final binary / plugin.
> 
> Hopefully this gets you started. Let me know if you get stuck anywhere.

hello and thank you for your interest. But I don't understand how to install it on my Mint system....

**Comment by CodesoundR** [comment]
> I drove this a bit further and opened Pull Requests in both project components:
> 
> #85 [tiborvass/libonnxruntime-neuralnote#2](https://github.com/tiborvass/libonnxruntime-neuralnote/pull/2)
> 
> However, communication quickly died down (or didn't happen), so I assumed that the maintainers were not interested in this. I finished the packaging for the Linux distribution of my choice (NixOS) and carried on. I'll give a short rundown on how to build this yourself, though. You will need to use my forks of [libonnxruntime-neuralnote](https://github.com/polygon/libonnxruntime-neuralnote) and [NeuralNote](https://github.com/polygon/NeuralNote). For both, you want the "Linux" branch.
> 
> I will link my Nix build recipes, since they are also a good indicator on the dependencies that you will need. You can ignore a lot of the CMake download links replacement, that is a Nix/NixOS peculiarity (no internet access during builds).
> # libonnxruntime-neuralnote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/libonnxruntime-neuralnote.nix))
> 
> Once you installed everything unter [`buildInputs`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L32-L41), you just run the instructions under [`buildPhase` ](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L52-L55). This should yield you a `libonnxruntime-neuralnote.tar.gz` which you will need in the next step.
> # NeuralNote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/neuralnote.nix))
> 
> Again, make sure you have all the [tools and dependencies](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L42-L58) installed. Now, run all the steps under [`postPatch`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L77-L86) in the root-folder of the repo (in line 83, reference the `tar.gz` file from the build above). This also patches a Drag&Drop bug in JUCE. You then just run CMake like [above](https://github.com/DamRsn/NeuralNote/issues/33#issuecomment-1784055326) and this should give you the final binary / plugin.
> 
> Hopefully this gets you started. Let me know if you get stuck anywhere.

hello and thank you for your interest. But I don't understand how to install it on my Mint system....

**Comment by Ken-Andre** [comment]
Ok, thanks for this contribution.

**Comment by polygon** [comment]
> hello and thank you for your interest. But I don't understand how to install it on my Mint system....

These instructions are to build the project, distribution for specific Linux systems comes later. I'm very sorry, but I don't have the time to build and maintain packages for a Linux distribution I am not using myself. If you find someone who knows how to package for Ubuntu/Mint and is interested, I'm happy to assist getting things started.

**Comment by CodesoundR** [comment]
> > hello and thank you for your interest. But I don't understand how to install it on my Mint system....
> 
> These instructions are to build the project, distribution for specific Linux systems comes later. I'm very sorry, but I don't have the time to build and maintain packages for a Linux distribution I am not using myself. If you find someone who knows how to package for Ubuntu/Mint and is interested, I'm happy to assist getting things started.

Hi,
I wanted to thank you anyway for what you did. I try to see in the various formums if there is someone who can help us for this porting.... bye

**Comment by karimdjemai** [comment]
Hi,
thank you so much for your work @polygon, 
I got it to build on archlinux but when I run it, I get the following error:

./build/NeuralNote_artefacts/Release/Standalone/NeuralNote 
terminate called after throwing an instance of 'Ort::Exception'
  what():  inference_session.cc:1020 LoadOrtModelWithLoader The ORT format model version [6] is not supported in this build 1.14.1. This build doesn't support ORT format models older than version 5. See: https://github.com/microsoft/onnxruntime/blob/rel-1.14.0/docs/ORT_Format_Update_in_1.13.mdlol 1
[1]    71338 IOT instruction (core dumped)  build/NeuralNote_artefacts/Release/Standalone/NeuralNote

which is particularily odd, since the newst ORT model version is 5 according to https://onnxruntime.ai/docs/performance/model-optimizations/ort-format-models.html

I did run the convert-onnx-to-ort stuff before building the onnxruntime and cp-ing the tar, but my guess is now that something went wrong in that step.
However I did not find a solution and have no idea on where to look further.
Does anybody have an idea on how to debug this? 
Thanks in advance!

**Comment by DamRsn** [comment]
I encountered a similar error message on Windows recently when working with onnxruntime. The issue was that my executable was linking with the onnxruntime dynamic library available in my system dynamic library directory. Maybe the same is happening for you on Linux.

So check that you successfully get a static lib for onnxruntime and that you link with it and not with some other version available on your system.

**Comment by karimdjemai** [comment]
Hi @DamRsn,
thanks for your reply!

I checked the dynamic library links of the executable, and onnxruntime was not there. The onnxruntimelib.a file sits in the right location as well. 

But you might still be onto something, as I could not find a log proving the linking of the static lib in the build output. This might be normal, though.
I did not find any other way to verify that the static lib file was linked correctly, do you have any idea?

I am very grateful for further ideas on where to look to get it running in general, thanks in advance!

**Comment by karimdjemai** [comment]
I finally got it working, so if anyone encounters my problem, here is how I solved it: 
In general I followed these steps outlined by @polygon

> I drove this a bit further and opened Pull Requests in both project components:
> 
> #85 [tiborvass/libonnxruntime-neuralnote#2](https://github.com/tiborvass/libonnxruntime-neuralnote/pull/2)
> 
> However, communication quickly died down (or didn't happen), so I assumed that the maintainers were not interested in this. I finished the packaging for the Linux distribution of my choice (NixOS) and carried on. I'll give a short rundown on how to build this yourself, though. You will need to use my forks of [libonnxruntime-neuralnote](https://github.com/polygon/libonnxruntime-neuralnote) and [NeuralNote](https://github.com/polygon/NeuralNote). For both, you want the "Linux" branch.
> 
> I will link my Nix build recipes, since they are also a good indicator on the dependencies that you will need. You can ignore a lot of the CMake download links replacement, that is a Nix/NixOS peculiarity (no internet access during builds).
> # libonnxruntime-neuralnote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/libonnxruntime-neuralnote.nix))
> 
> Once you installed everything unter [`buildInputs`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L32-L41), you just run the instructions under [`buildPhase` ](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/libonnxruntime-neuralnote.nix#L52-L55). This should yield you a `libonnxruntime-neuralnote.tar.gz` which you will need in the next step.
> # NeuralNote ([Nix-recipe](https://github.com/polygon/audio.nix/blob/master/vst/neuralnote/neuralnote.nix))
> 
> Again, make sure you have all the [tools and dependencies](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L42-L58) installed. Now, run all the steps under [`postPatch`](https://github.com/polygon/audio.nix/blob/1217ff4ed70c562b5eaf9832151967b3a568c65c/vst/neuralnote/neuralnote.nix#L77-L86) in the root-folder of the repo (in line 83, reference the `tar.gz` file from the build above). This also patches a Drag&Drop bug in JUCE. You then just run CMake like [above](https://github.com/DamRsn/NeuralNote/issues/33#issuecomment-1784055326) and this should give you the final binary / plugin.
> 
> Hopefully this gets you started. Let me know if you get stuck anywhere.


BUT when building  [libonnxruntime-neuralnote](https://github.com/polygon/libonnxruntime-neuralnote), you need to install the requirements.txt into the venv. Here it is important to use the correct python package version of onnxruntime (its 1.14.1) and not the newest. For that I had to downgrade python to below 3.11, so that the onnxruntime package is available.

@DamRsn Thanks for your help, without your comment I wouldn't have suspected that my build of  [libonnxruntime-neuralnote](https://github.com/polygon/libonnxruntime-neuralnote) was faulty, also thanks for building this incredible plugin in general. It is super cool!

**Comment by CodesoundR** [comment]
I hope to see soon an official version of this software for Linux because it is difficult for me to build it.... :-(
anyway, thank you for all.....

**Comment by DamRsn** [comment]
It took some time but it's finally here! I just released [NeuralNote v1.1.0](https://github.com/DamRsn/NeuralNote/releases/tag/v1.1.0), it includes Standalone and VST3 for Linux! 

Let me know if it works as expected. If so, I'll close this issue.

**Comment by gnac** [comment]
One question. I downloaded the pre-built vst and standalone, and was able to transcode an audio segment.
However, When I try to drag the midi file somewhere, eg file explorer or DAW it doesn't appear to do anything.
![image](https://github.com/user-attachments/assets/df83f9ef-be0d-4132-bcad-dbbe723728c2)
I'll try compiling locally to see if behavior changes.

**Comment by DamRsn** [comment]
Hey @gnac , thanks for reporting this.

What's your Linux distro and what DAW are you using? Is it both on Standalone and VST3?
It was working on my end (I'm on Ubuntu and used Reaper). I'll double check tomorrow.

If you're able to compile NeuralNote, testing with a debugger to see what happens when you try to drag would be great. You could put breakpoints in `MidiFileDrag.cpp`.

**Comment by gnac** [comment]
I use Manjaro (Arch deriviative) with KDE/Plasma and Bitwig. I see the same behavior in standalone and VST.
I also tested the VST in Carla.
Its probably worth noting that the drag and drop does not work on the audio load file either, but if you click on audio load file, you can navigate to the desired file.

**Comment by gnac** [comment]
I tried to build master, but I got `cp: cannot stat '../libonnxruntime-neuralnote/onnxruntime-v1.14.1-neuralnote.0-linux-x86_64.tar.gz': No such file or directory` when running build.sh
I switched to v0.1.0 and build.sh got as far as 
```
In file included from /home/dwh/AudioTools/src/NeuralNote.broken/Tests/Tests.cpp:2:
/home/dwh/AudioTools/src/NeuralNote.broken/Tests/../Lib/Model/Features.h:9:10: fatal error: onnxruntime_cxx_api.h: No such file or directory
    9 | #include <onnxruntime_cxx_api.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by gnac** [comment]
Instructions for building on Manjaro/Arch.
Arch Dependencies (install with pacman): gcc pyenv

# Build libonnxruntime
- Manually Download https://github.com/tiborvass/libonnxruntime-neuralnote/archive/refs/tags/v1.14.1-neuralnote.2.tar.gz
- extract the downloaded file to a folder called `libonnxruntime-neuralnote`. This is important so that the NeuralNote build can find the library artifacts later.
- Use Pyenv to create a python 3.10 environment, then follow the readme guide for python venv, eg.

 ```
pyenv shell 3.10.14 # specific to Arch, need to be run on each new console.
python3 -m venv venv # from readme
source ./venv/bin/activate # you have to do this everytime you re-enter a console environment for this build.
pip install -r requirements.txt  # one time requirement for this venv.
```
 
- Follow instructions for building libonx, eg `python3 -m venv venv` etc  `./release.sh v1.14.1-neuralnote.0`
- The first attempt will likely fail on the test activation file. 
```/home/dwh/AudioTools/src/libonnxruntime-neuralnote-1.14.1-neuralnote.2/onnxruntime/onnxruntime/test/mlas/unittest/test_activation.cpp:71:21: error: ‘setw’ is not a member of ‘std’
   71 |             << std::setw(8) << std::setfill('0') <<std::hex << Buffer[i].u << ", expecting:"
``
I haven't figured out why std::setw and std::setfill fail in this environment, but comment out lines 61-97 of `onnxruntime/onnxruntime/test/mlas/unittest/test_activation.cpp`.
Then, to avoid the release script deleteing and re-cloneing the onx repo because we've modified it, comment out lines 33-36 in release.sh
```
# if test -n "$(git -C ./onnxruntime status --porcelain)"; then
# 	rm -rf onnxruntime
# 	clone
# fi
```
Then, at line 49 in release.sh, I had to add the following before line 49 `mkdir -p "./$dir/include"`
so that the following `cp` line succeeds. eg lines 48-54 should now look like this:
```
(
	mkdir -p "./$dir/include"
	cp -rf include "./$dir/include"
	cp model.with_runtime_opt.ort "./$dir/"
	tar -czf "onnxruntime-${version}-${os}-${arch}.tar.gz" "$dir"
	rm -rf "./$dir/include" "./$dir/model.with_runtime_opt.ort"
)
```
Rerun ` ./release.sh v1.14.1-neuralnote.0`
 Now, you should have a `onnxruntime-v1.14.1-neuralnote.0-linux-x86_64.tar.gz` file in the `libonnxruntime-neuralnote` directory that NeuralNote needs.
 
 # Build NeuralNote
 Clone the NeuralNote repo to the parent of the onnxruntime folder.
`git clone --recurse-submodules --shallow-submodules https://github.com/DamRsn/NeuralNote`
enter the repo and checkout the linux compatible version.
`cd NeuralNote && git checkout v1.1.0`
Then build NeuralNote per the readme.
`build.sh`
However, at this point on my system, I get the following:
```
[ 81%] Built target NeuralNote
make[2]: *** No rule to make target '/home/dwh/AudioTools/src/NeuralNote/ThirdParty/onnxruntime/lib/libonnxruntime.a', needed by 'NeuralNote_artefacts/Release/VST3/NeuralNote.vst3/Contents/x86_64-linux/NeuralNote.so'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/home/dwh/AudioTools/src/NeuralNote/ThirdParty/onnxruntime/lib/libonnxruntime.a', needed by 'NeuralNote_artefacts/Release/Standalone/NeuralNote'.  Stop.
```

**Comment by gnac** [comment]
> I was debugging this for a while

What permutations did you have to go through to get this to build? I recognize I'm on a slightly different platform, but the build step for linux seem broken or at least incomplete in their current form.
I finally was able to get a static libonnxruntime.a built, and then I had to force it into a tarball to be used by the NeuralNote build, which finally allowed me to build. Just wondering if you had to go through the same process.

**Comment by gnac** [comment]
update 2: I just made a [pull request](https://github.com/tiborvass/libonnxruntime-neuralnote/pull/3) to the libonnxruntime-neuralnote repo to automatically create and package the libonnxruntime.a file needed by NeuralNote into the onnxruntime-v1.14.1-neuralnote.0-linux-x86_64.tar.gz tar ball.
Now that I've got a working NeuralNote build, I can do some testing/debugging.

**Comment by DamRsn** [comment]
> update 2: I just made a [pull request](https://github.com/tiborvass/libonnxruntime-neuralnote/pull/3) to the libonnxruntime-neuralnote repo to automatically create and package the libonnxruntime.a file needed by NeuralNote into the onnxruntime-v1.14.1-neuralnote.0-linux-x86_64.tar.gz tar ball. Now that I've got a working NeuralNote build, I can do some testing/debugging.

How is this PR different from [this one](https://github.com/tiborvass/libonnxruntime-neuralnote/pull/2)? That's the on I used to build on Linux and I think I will merge it soon. 

Concerning the drag and drop, is it solved? If not, I don't want to modify JUCE for this. What we could do instead is to add a button to export the midi file. It would opens a filechooser to select where to save the transcription.

**Comment by gnac** [comment]
I think having a filechooser to save the midi file is a reasonable workaround for now.

I hadn't seen the other pull request prior so it looks like the static library creation is the effectively the same with a some additional sub libraries added to the final. One minor difference is that the make-archive.sh file currently builds the archive using git commands, whereas the build-linux script in this pull request also builds an archive. this build-linux uses an arch setting which I don't see in my build, eg `/onnxruntime/build/linux_${arch}/${CMAKE_BUILD_TYPE}/` vs ` ./onnxruntime/build/Linux/${CMAKE_BUILD_TYPE}/`

**Comment by David-Else** [comment]
> Concerning the drag and drop, is it solved? If not, I don't want to modify JUCE for this. What we could do instead is to add a button to export the midi file. It would opens a filechooser to select where to save the transcription.

Hi, I am getting the same problem in Reaper on Linux, please add the export midi button soon, I would love to use this software :) 

![Image](https://github.com/user-attachments/assets/a2b7be44-0c1f-4548-a18d-68b70450cdae)

There is some talk that the standalone version drag and drop works better here: https://linuxmusicians.com/viewtopic.php?t=26632&sid=896ce3ee3602cfa9a4356ef9ecf52f32&start=15

**Comment by spectrumIG** [comment]
I'm sorry to bother but on Linux with Bitwig this plugin is unfortunatly not usable or at least is close to unusable. Does the PR opened since February resolve the issue?  Thanks and sorry if the tone seems harsh, it's really not my intention. Thanks again

**Comment by Anyeos** [comment]
Mine fails with a 403 http error. What are the script trying to download? 
I guess it is because cloudflare need to verify I am an human and it cannot be accomplished in a script.

`    error: downloading 'https://gitlab.com/libeigen/eigen/-/archive/d10b27fe37736d2944630ecd7557cefa95cf87c9/eigen-d10b27fe37736d2944630ecd7557cefa95cf87c9.zip' failed
`

**Comment by truong264vu** [comment]
Disclosure first: I build RiffToTab, a browser tool on the same basic-pitch model NeuralNote uses.

For anyone stuck on Linux while the build issues above get sorted: it runs in Chrome or Firefox with nothing to install, and it has the "Export MIDI" button this thread asks for, which downloads a .mid (with the tempo written in) instead of relying on drag-and-drop out of the plugin. Raw detection is the same model, so expect the same notes; the difference is only what is around it (browser, file export, a piano roll to edit in).

https://rifftotab.com/studio

Not a replacement for a native plugin in your DAW, but it may unblock someone today. Thanks DamRsn for NeuralNote, it is what showed the workflow works.


<img width="1417" height="831" alt="Image" src="https://github.com/user-attachments/assets/71e1dba0-0930-472a-a613-e6dbb1ce099f" />
